### PR TITLE
feat: index ERC20Votes

### LINKED ISSUE #3: feat: index ERC20-Votes proxy delegates
We should index ERC20-Votes deployment within this subgraph. With this it will be possible for a space to see their delegates dashboard directly, without having to setup an indexer. Here is some example on how adding a factory is done: https://github.com/snapshot-labs/sx-monorepo/blob/59435fbd72922b8babf2d4f3faeae05b4e3ebbe7/apps/subgraph-api/src/mapping.ts#L70-L74

### PR DESCRIPTION
**Pull request by pscott**
Adds indexing of the [GeneralPurposeFactory](https://etherscan.io/address/0x0f77c58bb8a75ed393123f9047e1787db637b251) and the specific [GenericERC20Votes](https://github.com/snapshot-labs/sx-periphery/blob/main/evm/src/GenericERC20Votes.sol) token.

Adds a `datasource` for the factory and a `template` for the token. The token uses the already defined `handleDelegateChanged` and `handleDelegateVotesChanged` event handlers.

An example "testing" contract has been deployed at [https://etherscan.io/address/0xFf162931b33F7C7bEE8C52006Db3df2f7ba7fDA7](https://etherscan.io/address/0xFf162931b33F7C7bEE8C52006Db3df2f7ba7fDA7).

I've delegated `100` units to myself in [this transaction](https://etherscan.io/tx/0x920e6075e0924ead37711bb6b8257c8a0997ca7daf4204814f3a53e6961699d7). Careful though, the total supply is `100`, but not `100 * 10**18` even though decimals is set to 18. This means that my amount might not be correctly displayed ^^ 

Closes #3

### CODE REVIEW

File: abis/GeneralPurposeFactory.json [hunk 1]
```diff
@@ -0,0 +1 @@
..."ds-test/=lib/openzeppelin-contracts-upgradeable/lib/forge-std/lib/ds-test/src/","erc4626-tests/=lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests/","forge-std/=lib/forge-std/src/","openzeppelin-contracts-upgradeable/=lib/openzeppelin-contracts-upgradeable/","openzeppelin-contracts/=lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/"],"optimizer":{"enabled":true,"runs":200},"metadata":{"bytecodeHash":"ipfs"},"compilationTarget":{"src/GeneralPurposeFactory.sol":"GeneralPurposeFactory"},"evmVersion":"paris","libraries":{}},"sources":{"lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/proxy/Clones.sol":{"keccak256":"0xd18408af8a91bedb3d56343eeb9b30eb852e6dea93a5e2d5c6db9ca4cb905155","urls":["bzz-raw://061475b7302a732dc598907790f407417551c2b13e89daa57b3698489ef5484b","dweb:/ipfs/QmdX5qmk3VSniSErFuD4aVawQxUi2MuUbP7spPcHddPfcX"],"license":"MIT"},"src/GeneralPurposeFactory.sol":{"keccak256":"0x046299eb611b932c32cdfc6804c3ab00e00338c0469c43b379e4a547e6ad48e4","urls":["bzz-raw://b4a66772187eb01c1ca745ae01f4be401e542fd7f549970bd584bfd79cfe043d","dweb:/ipfs/QmUGUx2yh2TAx499QfySwFeGmpEjcnrQjmfthi9ndKFPQA"],"license":"MIT"}},"version":1},"id":71}
```
Comment (bonustrack) [inline abis/GeneralPurposeFactory.json]: Can you only add the ABI, and have it formated like this file: https://github.com/snapshot-labs/delegates-subgraph/blob/6718011351028be93fafa46716297c6721215ac2/abis/Token.json

**Comment by bonustrack** [comment]
Did you tested sync a subgraph with it?

**Comment by pscott** [comment]
> Did you tested sync a subgraph with it?

No that's one the questions I had. How do I go on and test this? Do we have a testing environment? Access to a testing account?

File: abis/GeneralPurposeFactory.json (hunk 1 above)
Comment (pscott) [reply abis/GeneralPurposeFactory.json]: 0f5687955ecedcfd970161fd9c578daa00e2c091

**Comment by bonustrack** [comment]
@pscott There is way to run this locally but the easiest is probably to deploy it using your account on TheGraph

**Comment by pscott** [comment]
Previous deployment is invalid, as I'm using a new `implem` address now. Unfortunately not the best time to create a contract on ethereum atm (~$100). Will wait until we see gwei at < 10.

**Comment by pscott** [comment]
Tested on https://etherscan.io/address/0x8b958736f1a02839de9eb900b6e4ea4e935a4f34 . Works :D

**Comment by bonustrack** [comment]
Please try subgraph deployment on your own, and if it works we can merge this.

**Comment by pscott** [comment]
Tested, works

Screenshot 2024-09-30 at 16 59 45

File: src/mapping.ts [hunk 2]
```diff
@@ -1,8 +1,14 @@
-import { dataSource } from '@graphprotocol/graph-ts';
-import { DelegateChanged, DelegateVotesChanged } from '../generated/Token/Token';
+import { Address, dataSource } from '@graphprotocol/graph-ts';
+import { DelegateChanged, DelegateVotesChanged } from '../generated/templates/GenericERC20Votes/Token';
 import { toDecimal, getDelegate, getGovernance, BIGINT_ZERO } from './helpers';
+import { ContractDeployed } from "../generated/GeneralPurposeFactory/GeneralPurposeFactory";
```
Comment (bonustrack) [inline src/mapping.ts]: Can you use single quote everywhere?
```suggestion
import { ContractDeployed } from '../generated/GeneralPurposeFactory/GeneralPurposeFactory';
```

File: src/mapping.ts [hunk 3]
```diff
@@ -1,8 +1,14 @@
...
+import { GenericERC20Votes } from "../generated/templates";
+import { log } from '@graphprotocol/graph-ts';
```
Comment (bonustrack) [inline src/mapping.ts]: This pkg is already imported earlier but anyhow you should remove the log from the code.
```suggestion
```

Comment (pscott) [reply src/mapping.ts]: Ok sorry, thought I'd mimic what we have on [delegates-api](https://github.com/snapshot-labs/delegates-api/blob/d9059145960769c6da502b25412119059c15a63a/src/writer.ts#L10)

**Comment by bonustrack** [comment]
Looks good thanks, I've deployed the subgraph, will merge once it's fully sync https://thegraph.com/studio/subgraph/delegates/

**Pull request merged**